metricbeat/internal/sysinit/init.go

```diff
@@ -0,0 +1,112 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sysinit
+
+import (
+	"flag"
+	"sync"
+
+	"github.com/elastic/beats/v7/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/v7/libbeat/common/fleetmode"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+var hostFS = flag.String("system.hostfs", "", "Mount point of the host's filesystem for use in monitoring a host from within a container")
+
+var once sync.Once
+
+// A wrapper library that allows us to deal with the more complex HostFS setter logic required for legacy metricbeat code.
+// This will serve as a generic init function for either the system or linux module.
+
+// HostFSConfig is a bare struct for unpacking the config we get from agent or metricbeat
+type HostFSConfig struct {
+	HostFS string `config:"hostfs"`
+}
+
+// MetricbeatHostFSConfig
+type MetricbeatHostFSConfig struct {
+	HostFS string `config:"system.hostfs"`
+}
+
+// Init either the system or linux module. This will produce different modules depending on if we're running under agent or not.
+func InitSystemModule(base mb.BaseModule) (mb.Module, error) {
+	// common code for the base use case of `hostfs` being set at the module-level
+	logger := logp.L()
+	hostfs, userSet := findConfigValue(base)
+	if fleetmode.Enabled() {
+		logger.Infof("initializing HostFS values under agent: %s", hostfs)
+		return fleetInit(base, hostfs, userSet)
+	}
+	return metricbeatInit(base, hostfs, userSet)
+}
+
+func fleetInit(base mb.BaseModule, modulepath string, moduleSet bool) (mb.Module, error) {
+	once.Do(func() {
+		InitModule(modulepath)
+	})
+
+	// The multiple invocations here might seem buggy, but we're dealing with a case were agent's config schemea (local, per-datastream) must mesh with the global HostFS scheme used by some libraries
+	// Strictly speaking, we can't guarantee that agent will send consistent HostFS config values across all datastreams, as it treats a global value as per-datastream.
+	if moduleSet {
+		InitModule(modulepath)
+	}
+
+	return &Module{BaseModule: base, HostFS: modulepath, UserSetHostFS: moduleSet}, nil
+}
+
+// Deal with the legacy configs available to metricbeat
+func metricbeatInit(base mb.BaseModule, modulePath string, moduleSet bool) (mb.Module, error) {
+	var hostfs = modulePath
+	var userSet bool
+	// allow the CLI to override other settings
+	if hostFS != nil && *hostFS != "" {
+		cfgwarn.Deprecate("x", "The --system.hostfs flag will be removed in the future and replaced by a config value.")
```

Comment:
I think it is ok if you keep `8.0.0` here, even if we are not removing it on this version.
```suggestion
		cfgwarn.Deprecate("8.0.0", "The --system.hostfs flag will be removed in the future and replaced by a config value.")
```